### ISSUE TITLE
dataclass class literal subtyping with callables not working

### ISSUE THREAD

**Issue by MatthewMckee4** [astral-sh/ty#760]
### Summary

This snippet does not work
```py
@dataclass
class DC:
    x: DC | None

v: Callable[[DC], DC] = DC
```

The reason is that `__init__` for `DC` is not a `FunctionLiteral`.

Should we allow other types for these constructor functions too? (other than FunctionLiteral and Callable

**Comment by erictraut** [comment]
You may find [this section of the typing spec](https://typing.python.org/en/latest/spec/constructors.html#converting-a-constructor-to-callable) to be useful here. It covers the rules for converting the constructor of a class object into a callable type for purposes of assignability testing.

**Comment by MatthewMckee4** [comment]
> 2. If the class defines a `__new__` method

Does this mean we should not allow different types of `__new__`?

And same for `__init__`

**Comment by carljm** [comment]
We already implement that section of the spec, it's just that our implementation assumes that `__new__`/ `__init__` will be a `FunctionLiteral` type, and the synthesized `__init__` on a dataclass is not. I think ideally we should support any callable type as `__init__` or `__new__`, using a generalized `Type::into_callable_type` method. (But practically speaking, just adding support for `Type::Callable` so we support dataclass synthesized methods would go a long way.)

**Comment by MatthewMckee4** [comment]
Yeah, I'm trying to generalise it to use the into_callable method we just landed

**Comment by MatthewMckee4** [comment]
Just going for supporting 'Callable' as using into_callable has proved to be a lot more work
